miniapps/dpg/acoustics.cpp

```diff
@@ -0,0 +1,702 @@
+// -(u , ∇ q) + i ω (p , q) + < û, q >  = (f,q)   ∀ q ∈ H¹(Ω)
+//                                   p̂  = p₀      on ∂Ω
+
+// Note:
+// p̂ := p on Γₕ  (skeleton)
+// û := u on Γₕ
+
+// -------------------------------------------------------------
+// |   |     p     |     u     |    p̂      |    û    |  RHS    |
+// -------------------------------------------------------------
+// | v | -(p, ∇⋅v) | i ω (u,v) | < p̂, v⋅n> |         |         |
+// |   |           |           |           |         |         |
+// | q | i ω (p,q) |-(u , ∇ q) |           | < û,q > |  (f,q)  |
+
+// where (q,v) ∈  H¹(Ω) × H(div,Ω)
+
+// Here we use the "Adjoint Graph" norm on the test space i.e.,
+// ||(q,v)||²ᵥ = ||A^*(q,v)||² + ||(q,v)||² where A is the
+// acoustics operator defined by (1)
+
+#include "mfem.hpp"
+#include "util/complexweakform.hpp"
+#include "../common/mfem-common.hpp"
+#include <fstream>
+#include <iostream>
+
+using namespace std;
+using namespace mfem;
+using namespace mfem::common;
+
+void acoustics_solution(const Vector & X, complex<double> & p,
+                        vector<complex<double>> &dp, complex<double> & d2p);
+
+void acoustics_solution_r(const Vector & X, double & p,
+                          Vector &dp, double & d2p);
+
+void acoustics_solution_i(const Vector & X, double & p,
+                          Vector &dp, double & d2p);
+
+double p_exact_r(const Vector &x);
+double p_exact_i(const Vector &x);
+void u_exact_r(const Vector &x, Vector & u);
+void u_exact_i(const Vector &x, Vector & u);
+double rhs_func_r(const Vector &x);
+double rhs_func_i(const Vector &x);
+void gradp_exact_r(const Vector &x, Vector &gradu);
+void gradp_exact_i(const Vector &x, Vector &gradu);
+double divu_exact_r(const Vector &x);
+double divu_exact_i(const Vector &x);
+double d2_exact_r(const Vector &x);
+double d2_exact_i(const Vector &x);
+double hatp_exact_r(const Vector & X);
+double hatp_exact_i(const Vector & X);
+void hatu_exact(const Vector & X, Vector & hatu);
+void hatu_exact_r(const Vector & X, Vector & hatu);
+void hatu_exact_i(const Vector & X, Vector & hatu);
+
+int dim;
+double omega;
+
+enum prob_type
+{
+   plane_wave,
+   gaussian_beam
+};
+
+prob_type prob;
+
+int main(int argc, char *argv[])
+{
+   const char *mesh_file = "../../data/inline-quad.mesh";
+   int order = 1;
+   int delta_order = 1;
+   bool visualization = true;
+   double rnum=1.0;
+   int ref = 0;
+   bool static_cond = false;
+   int iprob = 0;
+
+   OptionsParser args(argc, argv);
+   args.AddOption(&mesh_file, "-m", "--mesh",
+                  "Mesh file to use.");
+   args.AddOption(&order, "-o", "--order",
+                  "Finite element order (polynomial degree)");
+   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
+                  "--no-visualization",
+                  "Enable or disable GLVis visualization.");
+   args.AddOption(&rnum, "-rnum", "--number_of_wavelenths",
```

Comment:
The typo `wavelenths` is in several files.

Reply:
This is indeed inexcusable 😄